Renamed the max_round parameter

pascalaubry opened the pull request:
The aim of this PR is to make ranking code more understandable by everybody:
- Renamed ranking methods *(max_round) to *\_after_round(round__) or *\_before_round(round\_) or *\_at_round(round\_)
- Removed the tournament parameter where it was not needed

Molrn committed:
Adapt tie-breaks to plugin architecture

Molrn committed:
Adapt tie-breaks to plugin architecture

Molrn committed:
remove individual.py

Molrn committed:
remove individual.py

Molrn committed:
comments

Molrn committed:
comments

Molrn committed:
rename ffe tie_break file

Molrn committed:
rename ffe tie_break file

pascalaubry committed:
Renamed ranking methods *(max_round) to *_after_round(round_) or *_before_round(round_) and removed the tournament parameter where it was not needed

pascalaubry commented:
Note: this PR is based on @Molrn 's branch plugin-tie-breaks, it should be reviewed after PR #242 has been merged to dev.

pascalaubry committed:
Merge branch 'plugin-tie-breaks' into rename-max-round

pascalaubry committed:
Made the tie-break tests work without passing the tournament

pascalaubry committed:
Fixed FFE get_extra_tie_break_classes()

pascalaubry committed:
Fixed FFE get_extra_tie_break_classes()

pascalaubry committed:
Merge branch 'plugin-tie-breaks' into rename-max-round

pascalaubry committed:
Merged plugin-tie-breaks

pascalaubry committed:
Merge branch 'dev' into rename-max-round

pascalaubry committed:
Make handicap string more understandable

Molrn commented on src/data/player.py:
```diff
@@ -131,15 +142,16 @@ def total_points(self, only_played: bool = False) -> float:
+    #        for round_index, pairing in self.pairings.items()
+    #        if round_index <= after_round and
+    #        (pairing.played or not only_played)
+    #    )
```
C'est dégueu du code commenté comme ça, pour moi il faut supprimer dans ces cas là. La probabilité que ce soit réutilisé ne vaut pas le fait que ça traine comme ça. quitte à laisser trainer "au cas ou", autant ne pas commenter la méhode

pascalaubry replied:
Already deleted.

Amaras commented on src/data/player.py:
```diff
@@ -131,15 +142,16 @@ def total_points(self, only_played: bool = False) -> float:
-            if round_index <= max_round and
-            (pairing.played or not only_played)
-        )
+    # unused method
+    #def max_possible_points(self, after_round: int | None = None, only_played: bool = False) -> float:
+    #    if after_round is None:
+    #        after_round = max(self.pairings)
+    #    return sum(
+    #        Result.GAIN.points(self.point_values)
+    #        for round_index, pairing in self.pairings.items()
+    #        if round_index <= after_round and
+    #        (pairing.played or not only_played)
+    #    )
```
you can remove it, we'll add it back if necessary

pascalaubry replied:
Done.

Amaras commented on src/data/player.py:
```diff
@@ -283,16 +286,15 @@ def rating_type(self) -> PlayerRatingType:
-        from round 1 to round `max_round` (returns None)"""
+    def compute_points_before_round(self, before_round: int):
+        """Computes and stores the points scored by the player before round `before_round` (returns None)"""
         # NOTE(Amaras) this does not rely on the fact that insertion order
         # is preserved in 3.6+ dict, because I can't be sure insertion order
         # is the correct (increasing) round order
         # NOTE(Amaras) if you were to include the current round
         # in the computation, boards regularly change their ordering
         # during the current round as results are added
```
This comment should be moved to `points_before_round` and `points_after_round` (or whatever the new name is)

pascalaubry replied:
Done.

Amaras commented on src/data/tie_break.py:
```diff
@@ -926,26 +920,26 @@ def acronym(self) -> str:
-            if round_index <= max_round and pairing.opponent_id is not None
-            and pairing.played
+            if round_index <= round_ and pairing.opponent_id is not None
+               and pairing.played
```
```suggestion
            and pairing.played
```

pascalaubry replied:
Done.

Molrn commented on src/data/player.py:
```diff
@@ -283,16 +286,15 @@ def rating_type(self) -> PlayerRatingType:
-    def compute_points(self, max_round: int):
-        """Computes and stores the points of the player,
-        from round 1 to round `max_round` (returns None)"""
+    def compute_points_before_round(self, before_round: int):
```
valid for all similar changes: naming the function and the parameter is too much in my opinion. I much prefer shorter function names : `compute_points` with `before_round` as a parameter

pascalaubry replied:
I do agree with that having both compute_points_before_round and before_round is too much but I would really prefer to keep the function name and renaming the parameter to round_.

pascalaubry replied:
Removed method suffixes _after and _before and renamed the round_ parameter.

Amaras commented on tests/test_tiebreaks.py:
```diff
@@ -351,7 +353,7 @@ def test_points(self):
         tie_break_ = tie_break.WinsTieBreak()
         results = {
-            player.id: tie_break_.compute_player_value(player, self.tournament)
+            player.id: tie_break_.compute_player_value_after_round(player)
```
That is confusing, I'd expect the round number to be a parameter.
You should add a method passing `round_=None`, instead of not passing anything there.

pascalaubry replied:
Indeed, I made the round_ parameter mandatory (practically most of the calls with no parameter were in test_tie_breaks.py).
Tests still successful:
Ran 36 tests in 0.068s
OK

Amaras reviewed (changes requested):
A few things feel weird, but good overall

Molrn commented on src/data/player.py:
```diff
@@ -311,8 +313,10 @@ def add_points(self, points: float):
 
-    def to_trf(
-        self, player_id_to_trf_id: Callable[[int], int], max_round: int,
+    def to_trf_after_round(
```
pareil: aucun besoin d'ajouter after_round. pour moi le cas utile ou tu fais une distinction `before` / `after` c'est quand t'as 2 fonctions similaires mais pour un usage différent (ex: player.points_before / player.points_after)

pascalaubry replied:
Well, clear to you because you wrote the method. Practically I had to read the code to know if if was before or after.
At least a comment is needed but IMO everything is clear after renaming the method.

Amaras replied:
`to_trf_after` could have worked, instead of adding yet another word.
But yeah, all methods need docstrings.
I'll take some time to do that work after 2.5.0 is out.

Molrn replied:
to_trf(after_roud) is very clear IMO. TRF is not even about rounds, these namings make all the functions about rounds. If it's an optional parameter in most of these it's because it's not the main focus. Renaming avery function after an optional parameter is not ok to me

pascalaubry replied:
I agree that
`def to_trf(player, round_after)`
is clear for the declaration of the method.
But when you call the method then
`player.to_trf(player, self.current_round + 1)`
is less clear than
`player.to_trf_after(player, self.current_round + 1)`
That is why I think that using `_after` or `_before` in the method name is easier.
Anyway not worth fighting more on this, I let you change back `to_trf_after` to `to_trf` I you prefer.

Amaras replied:
If you have an ambiguity problem, make the argument keyword only (add a `*` in the argument list before it), just make sure to touch all uses

Molrn replied:
borrowing a page from Sammy's playbook: `to_trf(player_id_to_trf_id, /, *, after_round: int | None = None)` forces the call to the function to use parameter naming. `to_trf(player_id_to_trf_id, 5)` is invalid, `to_trf(player_id_to_trf_id, after_round=5)` is good

pascalaubry replied:
Removed method suffixes _after and _before and renamed the round_ parameter.

Molrn commented on src/database/sqlite/event/event_database.py:
```diff
@@ -1551,7 +1551,7 @@ def _dump_to_json_database_tie_breaks(
     ) -> str | None:
         """Serializes the tie breaks into JSON. Returns a serialization
-        with format [{'type': str, 'options': {str: value}}]."""
+        with format [{'tie_break': str, 'options': {str: value}}]."""
```
```suggestion
        with format [{'type': str, 'options': {str: value}}]."""
```

pascalaubry replied:
Is it something I missed when merging?

Molrn replied:
I don't know, it semt like to merged on the right side at this moment for me, but apparently not.. The mystery of merging

pascalaubry replied:
OK that was just on the comment, fixed.

Molrn commented on src/data/tie_break.py:
```diff
@@ -157,11 +157,10 @@ def acronym(self) -> str:
+    def compute_player_value_after_round(
         self,
         player: 'Player',
-        tournament: 'Tournament',
```
a change I asked myself about: removing the tournament from the tie break computation interface and using the tournament ref. with tournament in the interface, we save the usage of the ref when bulk computing the tie-breaks for all the players of a tournament, but I don't know if this is event relevent. @Amaras I'd like your opinion on that

Amaras replied:
No qualms about that, but we need to make sure the tournament object is alive before using it

pascalaubry replied:
@Amaras You mean that the GC could flush the tournament during a request?

pascalaubry replied:
Checked on WhatsApp.

Molrn commented on src/data/player.py:
```diff
@@ -103,24 +104,34 @@ def __init__(
+
+    @property
+    def tournament(self) -> 'Tournament | None':
+        return self._tournament_ref() if self._tournament_ref else None
+
+    @tournament.setter
+    def tournament(self, tournament: 'Tournament | None'):
+        self._tournament_ref = weakref.ref(tournament) if tournament else None
 
     @property
     def point_values(self) -> dict[Result, float] | None:
         return self._point_values
 
-    def points_before(self, max_round: int, only_played: bool = False) -> float:
+    def points_before_round(self, round_: int, only_played: bool = False) -> float:
```
`points_before(round)` is already clear enough in my opinion. As a general rule, repeating parameter names in the function name is something to prohibit

pascalaubry replied:
@Molrn Better for you on the last commit?

Molrn replied:
answered on Whatsapp, here's the convention I suggest:  
- if round is an optional parameter, specify its usage by the naming of the parameter and the docstring. 
- If it is a mandatory parameter, specify before or after (or before_round / after_round) in the name of the function.

pascalaubry replied:
I had missed compute_player_ranks_after() and adjusted_score_after(), done for both methods.

pascalaubry committed:
Final review #245
Made the round_ parameter of compute_player_value_after_round(player, round_) mandatory

Molrn reviewed (commented):
I'm strongly opposed to some of the changes introduced in this PR. In general, changing the whole API and function naming is something to prohibit as it might introduce issues, so doing so for a change that is in my opinion very controversial is a big no for me.  

I agree that something can be done to clarify these distinct usages, but renaming variables to before_round, after_round, at_round and round_ is already enough for me. Another solution is through comments. Using function names for that is really using a bazooka to kill a fly for me.  

@Amaras I'd like you opinion on all this

pascalaubry committed:
Removed unused imports

pascalaubry committed:
Removed unused locale variable

pascalaubry committed:
Shortened the names of the methods and the names of the parameters

pascalaubry committed:
Fixed comment

Molrn commented on src/data/tournament.py:
```diff
@@ -580,9 +580,9 @@ def read_papi(self, update: bool = False):
+        self._calculate_points_before_current_round()
         self._build_boards()
-        self.estimate_players(papi_legacy=True)
+        self.estimate_players_after(papi_legacy=True)
```
round is not a parameter. You can add an optional parameter and specify the default in the docstring

pascalaubry replied:
Removed method suffixes _after and _before and renamed the round_ parameter.

Molrn commented on src/data/player.py:
```diff
@@ -311,8 +308,10 @@ def add_points(self, points: float):
+    def to_trf_after(
+        self,
+        player_id_to_trf_id: Callable[[int], int],
+        round_: int,
```
here there should be an after_round parameter should be optional, and to_trf_after-> to_trf

pascalaubry replied:
Removed method suffixes _after and _before and renamed the round_ parameter.

Molrn commented on src/data/player.py:
```diff
@@ -433,24 +432,25 @@ def set_time_control(self, initial_time: int, increment: int, modified: bool):
 
-    def compute_tie_break_values(
-        self, tournament: 'Tournament', max_round: int | None = None
+    def compute_tie_break_values_after(
```
round parameter optional: compute_tie_break_values(after_round)

pascalaubry replied:
Removed method suffixes _after and _before and renamed the round_ parameter.

Molrn commented on src/data/tie_break.py:
```diff
@@ -157,11 +157,10 @@ def acronym(self) -> str:
     @abstractmethod
-    def compute_player_value(
+    def compute_player_value_after(
         self,
```
compute_player_value(after_round)

pascalaubry replied:
Removed method suffixes _after and _before and renamed the round_ parameter.

Molrn commented on src/data/tie_break.py:
```diff
@@ -229,23 +228,25 @@ class TieBreakUtils:
-        games for the last round not determined over the board are considered as draws."""
+        games not already determined are considered a draw."""
+        tournament: 'Tournament' = player.tournament
+        if round_ is None:
+            round_ = max(player.pairings)
```
merge issue, this was removed
```suggestion
```

pascalaubry replied:
Fixed.

Molrn commented on src/data/tie_break.py:
```diff
@@ -229,23 +228,25 @@ class TieBreakUtils:
         Only adjusts them in case of requested byes followed by all VUR.
         If *adjust_fore* is True, the adjusted score for Fore Buchholz is computed:
-        games for the last round not determined over the board are considered as draws."""
+        games not already determined are considered a draw."""
```
```suggestion
        games for the last round not determined over the board are considered as draws."""
```
merge issue

pascalaubry replied:
Fixed.

Molrn commented on src/data/tie_break.py:
```diff
@@ -229,23 +228,25 @@ class TieBreakUtils:
         player: 'Player',
-        tournament: 'Tournament',
-        max_round: int,
+        round_: int | None = None,
```
adjusted_score(after_round: int)

pascalaubry replied:
Removed method suffixes _after and before and renamed the round parameter.

Molrn commented on src/data/tie_break.py:
```diff
@@ -268,16 +269,16 @@ def adjusted_score(
+    def buchholz_dummy_score_after(
         player: 'Player',
-        max_round: int = 1,
+        round_: int = 1,
```
buchholz_dummy_score(after_round)

pascalaubry replied:
Removed method suffixes _after and before and renamed the round parameter.

Molrn commented on src/data/tie_break.py:
```diff
@@ -1054,13 +1048,13 @@ def compute_player_value(
     @staticmethod
-    def _dummy_score(
+    def _dummy_score_after(
         player: 'Player',
```
_dummy_score

pascalaubry replied:
Removed method suffixes _after and before and renamed the round parameter.

pascalaubry committed:
Made the round_ parameter mandatory in compute_player_ranks_after() and adjusted_score_after()

Molrn commented on src/data/tournament.py:
```diff
@@ -692,12 +692,12 @@ def _calculate_player_virtual_points(
-        """Estimate the players after *max_round*.
-        If *max_round* is None, use the current round if possible.
+    def estimate_players_after(self, *, round_: int | None = None, papi_legacy: bool = True):
+        """Estimate the players after round *round_*.
```
estimate_player(after_round)

pascalaubry replied:
Removed method suffixes _after and _before and renamed the round_ parameter.

Molrn commented on src/data/tournament.py:
```diff
@@ -818,20 +818,22 @@ def _set_players_illegal_moves(self):
 
-    def compute_player_ranks(
-        self, max_round: int | None = None, papi_legacy: bool = True
+    def compute_player_ranks_after(
+        self,
```
compute_player_ranks(after_round)

pascalaubry replied:
Removed method suffixes _after and _before and renamed the round_ parameter.

Molrn commented on src/plugins/ffe/ffe_tie_break.py:
```diff
@@ -21,19 +21,19 @@ def papi_id(self) -> str:
 class AbstractPapiBuchholzTieBreak(AbstractPapiTieBreak, ABC):
     @staticmethod
-    def _papi_adjusted_score(
+    def _papi_adjusted_score_after(
```
_papi_adjusted_score(after_round)

pascalaubry replied:
Removed method suffixes _after and _before and renamed the round_ parameter.

Molrn commented on src/plugins/ffe/ffe_tie_break.py:
```diff
@@ -52,15 +52,15 @@ def _papi_adjusted_score(
 
     @staticmethod
-    def _papi_dummy_score(
+    def _papi_dummy_score_after(
```
_papy_dummy_score(after_round)

pascalaubry replied:
Removed method suffixes _after and _before and renamed the round_ parameter.

Molrn commented on src/plugins/ffe/ffe_tie_break.py:
```diff
@@ -79,32 +79,31 @@ def _papi_buchholz_cut(tournament_rounds: int) -> int:
         return 3
 
-    def compute_papi_buchholz_player_value(
+    def compute_papi_buchholz_player_value_after(
```
compute_papi_buchholz_player_value(after_round)

pascalaubry replied:
Removed method suffixes _after and _before and renamed the round_ parameter.

pascalaubry committed:
Removed method suffixes _after and _before
Renamed the round_ parameter

pascalaubry committed:
Fixed merge issue in tie_break.py

pascalaubry committed:
Merge branch 'dev' into rename-max-round

pascalaubry committed:
Renamed _calculate_player_virtual_points_after() to _calculate_player_virtual_points()

pascalaubry committed:
Fixed comments

Amaras commented on src/data/player.py:
```diff
@@ -103,24 +104,46 @@ def __init__(
+        # NOTE(Amaras) this does not rely on the fact that insertion order
+        # is preserved in 3.6+ dict, because I can't be sure insertion order
+        # is the correct (increasing) round order
+        # NOTE(Amaras) if you were to include the current round
+        # in the computation, boards regularly change their ordering
+        # during the current round as results are added
```
This part only matters for some purposes, I'll update comments and docstrings in a future PR

pascalaubry replied:
OK

Amaras commented on src/data/player.py:
```diff
@@ -283,16 +287,12 @@ def rating_type(self) -> PlayerRatingType:
-        # in the computation, boards regularly change their ordering
-        # during the current round as results are added
-        self.points = self.points_before(max_round)
+    def compute_points(
+            self,
+            before_round: int
+    ):
```
maybe this should be a keyword argument to avoid ambiguity
```suggestion
    def compute_points(
            self,
            *,
            before_round: int
    ):
```

Amaras commented on src/data/player.py:
```diff
@@ -312,7 +312,9 @@ def add_points(self, points: float):
 
     def to_trf(
-        self, player_id_to_trf_id: Callable[[int], int], max_round: int,
+        self,
+        player_id_to_trf_id: Callable[[int], int],
+        after_round: int,
```
Make `after_round` keyword-only to avoid ambiguity.
Maybe `player_id_to_trf_id` should be not be a positional-only argument, however.
```suggestion
        self,
        player_id_to_trf_id: Callable[[int], int],
        /,
        *,
        after_round: int,
```

Amaras commented on src/data/player.py:
```diff
@@ -441,16 +443,17 @@ def tie_break_values_as_strings(self) -> list[str]:
 
     def compute_tie_break_values(
-        self, tournament: 'Tournament', max_round: int | None = None
+        self,
+        after_round: int | None
```
```suggestion
        self,
        *,
        after_round: int | None
```

Amaras commented on src/data/tie_break.py:
```diff
@@ -160,8 +160,7 @@ def acronym(self) -> str:
     def compute_player_value(
         self,
         player: 'Player',
-        tournament: 'Tournament',
-        max_round: int | None = None,
+        after_round: int | None,
```
Make it keyword-only? (can't add a suggestion)

Amaras commented on src/data/tie_break.py:
```diff
@@ -231,21 +230,21 @@ class TieBreakUtils:
     def adjusted_score(
         player: 'Player',
-        tournament: 'Tournament',
-        max_round: int,
+        after_round: int | None,
         adjust_fore: bool = False,
```
Make the new arguments keyword-only

Amaras reviewed (dismissed):
What I said about `after_round` is valid for all instances.

pascalaubry committed:
Merge branch 'dev' into rename-max-round

pascalaubry committed:
Fixed comment

pascalaubry committed:
Made at_round, after_round and before_round keyword arguments

Molrn commented on src/data/player.py:
```diff
@@ -441,16 +454,18 @@ def tie_break_values_as_strings(self) -> list[str]:
     @property
     def rank(self) -> int:
-        assert self._rank, 'Player._rank is not set, call Tournament.compute_player_ranks() before.'
+        assert self._rank, 'Player._rank is not set, call Tournament.compute_player_ranks_after() before.'
```
```suggestion
        assert self._rank, 'Player._rank is not set, call Tournament.compute_player_ranks() before.'
```

Molrn commented on src/data/tie_break.py:
```diff
@@ -231,21 +231,22 @@ class TieBreakUtils:
-        tournament: 'Tournament',
-        max_round: int,
+        *,
+        after_round: int | None,
```
decided in https://github.com/papi-web-org/papi-web/pull/242#discussion_r1997538672 that the round parameter should not be None here.

Molrn commented on src/plugins/ffe/ffe_tie_break.py:
```diff
@@ -80,31 +82,31 @@ def _papi_buchholz_cut(tournament_rounds: int) -> int:
-        max_round: int | None,
-        use_cut_top: bool = False,
-        use_cut_btm: bool = False,
+            self,
+            player: Player,
+            *,
+            after_round: int | None,
+            use_cut_top: bool = False,
+            use_cut_btm: bool = False,
```
```suggestion
        self,
        player: Player,
        *,
        after_round: int | None,
        use_cut_top: bool = False,
        use_cut_btm: bool = False,
```

Molrn commented on src/plugins/ffe/ffe_tie_break.py:
```diff
@@ -155,13 +157,13 @@ def acronym(self) -> str:
-        player: 'Player',
-        tournament: 'Tournament',
-        max_round: int | None = None
+            self,
+            player: 'Player',
+            *,
+            after_round: int | None,
```
```suggestion
        self,
        player: 'Player',
        *,
        after_round: int | None,
```

Molrn commented on tests/test_tiebreaks.py:
```diff
@@ -588,7 +590,7 @@ def test_buchholz_cut1(self):
         results = {
             player.id: tie_break.TieBreakUtils.adjusted_score(
-                player, self.tournament, 5
+                player, after_round=None,
```
```suggestion
                player, after_round=self.tournament.rounds,
```
after_round has to be initialized for adjusted_score (tests are failing)

Molrn commented on tests/test_tiebreaks.py:
```diff
@@ -615,7 +617,7 @@ def test_adjusted_score(self):
         results = {
             player.id: tie_break.TieBreakUtils.adjusted_score(
-                player, self.tournament, 5, adjust_fore=True
+                player, after_round=None, adjust_fore=True
```
```suggestion
                player, after_round=self.tournament.rounds, adjust_fore=True
```
same here

Molrn reviewed (dismissed):
A few knit picks, and one fix because tests are failing. Otherwise, good for me ;)

pascalaubry committed:
Last fixes for #245

Molrn reviewed (approved):
LGTM
